quinn-proto/src/token_store.rs

```diff
@@ -0,0 +1,18 @@
+use bytes::Bytes;
+
+/// Responsible for storing address validation tokens received from servers and retrieving them for
+/// use in subsequent connections
+pub trait TokenStore: Send + Sync {
+    /// Potentially store a token for later one-time use
+    ///
+    /// Called when a NEW_TOKEN frame is received from the server.
+    fn insert(&self, server_name: &str, token: Bytes);
```

Comment:
I don't love using a trait to expose events to users. For example, it makes it a bit awkward to use async I/O to save them. Could we instead expose a new connection-level event at the proto level, and async getter on `Connection` at the quinn level? I admit this may make it more difficult to offer caching by default, but I'm not sure how useful a non-persistent cache is to most applications anyway. This would also be a breaking change. Thoughts?

Reply:
Resolving per below discussion: an async-friendly API can be gracefully added in future work.